### PR TITLE
Support variable do loop increment

### PR DESCRIPTION
**Pull request by Smit-create**
Trying to fix https://github.com/certik/minpack/commit/5ab467d37313d1c61e45b5a7bf94d167575b2685.
#512

### CODE REVIEW

File: src/libasr/pass/pass_utils.cpp [hunk 1]
```diff
@@ -632,15 +632,21 @@ namespace LFortran {
...
                     if (c->type == ASR::exprType::IntegerConstant) {
                         increment = ASR::down_cast<ASR::IntegerConstant_t>(c)->m_n;
                     } else if (c->type == ASR::exprType::IntegerUnaryMinus) {
                         ASR::IntegerUnaryMinus_t *u = ASR::down_cast<ASR::IntegerUnaryMinus_t>(c);
                         increment = - ASR::down_cast<ASR::IntegerConstant_t>(u->m_arg)->m_n;
                     } else {
-                        throw LCompilersException("Do loop increment type not supported");
+                        // This is the case when we don't have a constant value.
+                        // For now assuming that the increment is positive
+                        increment = 1;
```
Comment (certik) [inline src/libasr/pass/pass_utils.cpp:649]: I think this will not work if the increment `k` is 2, as in the example above, correct?

Comment (Smit-create) [reply src/libasr/pass/pass_utils.cpp:649]: It works. I have added a check that the increment should always be an integer. And if it is not a constant as in this case, we assume that increment will be positive.

Comment (Smit-create) [reply src/libasr/pass/pass_utils.cpp:649]: The test failures are because of `julia` backend which we need to modify a bit to make the tests pass.

Comment (Smit-create) [reply src/libasr/pass/pass_utils.cpp:649]: We can also make this more robust and reliable by adding an `if` statement that checks whether the `increment` is greater than zero and generates the `while` loop accordingly.

**Comment by Smit-create** [comment]
@certik @czgdp1807 I have updated this with a more robust fix that will work with any runtime value of increment. This could even work in LPython.

**Review by czgdp1807 — APPROVED** [review]
This looks fine to me.

**Review by certik — APPROVED** [review]
Awesome, this looks great! A very fast implementation for const increments, and a slower (runtime if statement) for non-constant increments. Thank you!

File: src/libasr/pass/pass_utils.cpp [hunk 2]
```diff
@@ -632,15 +632,43 @@ namespace LFortran {
...
+                            c, ASR::cmpopType::Gt, const_zero, log_type, nullptr));
+                        Vec<ASR::stmt_t*> if_body, or_body, result;
+                        if_body.reserve(al, 2);
+                        or_body.reserve(al, 2);
+                        result.reserve(al, 1);
+                        if_body = replace_doloop(al, loop, ASR::cmpopType::LtE);
+                        or_body = replace_doloop(al, loop, ASR::cmpopType::GtE);
+                        result.push_back(al, ASRUtils::STMT(ASR::make_If_t(al, loc, test, if_body.p,
+                                if_body.size(), or_body.p, or_body.size())));
+                        return result;
```
Comment (certik) [inline src/libasr/pass/pass_utils.cpp:671]: We can comment out this part, just put a note there that it currently fails if there are any "continue" statements.

Then underneath we can rework it to put the if statement inside the loop.

**Pull request merged**